### ISSUE TITLE
recording_stereo_driver.launch seems not working with an EVK1 and an EVK2

### ISSUE THREAD

**Issue by AndreasAZiegler** [berndpfrommer/metavision_ros_driver#13]
Me again. Towards my goal of having an EVK1 and an EVK2 synced, with trigger-in in my ROS setup, I have achieved the following:
* A custom capture software where sync and trigger-in works but not with ROS (Therefore, I know that the hardware is running properly)
* Running and recording the EVK1 with trigger-in with `recording_driver.launch`
* Running and recording the EVK2 with trigger-in with `recording_driver.launch`

Unfortunately, I failed to get `recording_stereo_driver.launch` working with my setup. After a bit digging in the code, it looks like the following.
* `MetavisionWrapper::cdCallback` is added as a callback [here](https://github.com/berndpfrommer/metavision_ros_driver/blob/master/src/metavision_wrapper.cpp#L319)
* But [the callback](https://github.com/berndpfrommer/metavision_ros_driver/blob/master/src/metavision_wrapper.cpp#L420) is never executed in my case

I do assume that I might have missed something. But I hope that @berndpfrommer might have some hints what I'm doing wrong, given that the single setups did work properly.

**Comment by berndpfrommer** [comment]
If the cdCallback is added (I assume you put a printout in there to make sure it is indeed added in your configuration) and the cdCallback does not happen then it could be some system issue. Meaning I have no clue what's going on and can only suggest some ugly high-level experiments to narrow this down:
Run two separate nodes, one for each camera, and forget about the recording for now. Run a straight node, not even a nodelet/composable. Don't turn on triggers, just sync. Can you get sync to work then, i.e. does the callback happen? If no, then does the callback happen when you start only one node? This procedure would tell us if it's some shared resource in the metavision driver that is blocking the callbacks. By running two separate nodes you eliminate the possibility of this being some interaction due to running in the same address space.
Basically the idea is to get sync to work in any possible configuration, then see what kills the callbacks.

**Comment by AndreasAZiegler** [comment]
I believe I found the issue. Currently how the sync mode with a `primary` and a `secondary` device is implemented seems to be the following:
* Secondary node starts camera
* Primary node waits for a ready signal
* Secondary node sets callback
* Because the primary camera is not started, the `cdCallback` of the secondary node is never called due to the missing sync signal

As a workaround, I moved [the start](https://github.com/berndpfrommer/metavision_ros_driver/blob/master/src/driver_ros1.cpp#L51) outside of the `primary`, `secondary` check so that both cameras are started. With the sync signal present after the `primary` camera is started, the `secondary` node does receive the `ready` signal.

**Comment by berndpfrommer** [comment]
You describe the startup-dance correctly, this is how it was meant to work, and it does work for a pair of SilkyEVCams.
There is a reason the primary is not started until the "ready" from the secondary is available: the time stamps will be off.
The question is: how does the secondary know what time stamps the primary puts on its packets? My understanding was that the sync cable does not transmit explicit time stamps, it just transmits a clock signal. But if it's just a clock signal, then there could be a constant shift in time stamps between the two cameras, right?
I believe this is how it was supposed to work: *before* the primary starts incrementing its time stamps, the secondary must be ready-and-waiting for the clock signal. The secondary sets its time stamp to zero and waits for the primary to send the clock signal. With every clock signal the secondary receives it bumps its time stamp by one unit (making this up). The primary also bumps its time stamp every time it sends a clock signal, so the two time stamps are in sync. This will only work though *if the secondary is ready and waiting* by the time the primary starts sending the clock signal. Hence the start-up dance.
If my assertion is correct, your time stamps should be synced but shifted. Not what we want.
There is a tool called "sync_test" in https://github.com/berndpfrommer/event_array_tools
You can use that to test if your timestamps are good.

**Comment by berndpfrommer** [comment]
Here is more about sync and ahh, they changed the behavior between Gen3 and Gen4!

https://docs.prophesee.ai/stable/hw/manuals/synchronization.html#chapter-synchronization

**Comment by berndpfrommer** [comment]
Can you try PR #15? This does the startup dance now without requiring events on the secondary to come.

**Comment by AndreasAZiegler** [comment]
Thanks for your work @berndpfrommer. I will continue working on this topic again next week and will provide you an update.

**Comment by AndreasAZiegler** [comment]
PR #15 seems to have solved the issue. Thanks again for the fast response.
